src/testpool.jl

```diff
@@ -146,40 +137,120 @@ resize_test_engine_pool!(10, id->"RAITest-test-\$id")
 resize_test_engine_pool!(0)
 ```
 """
-function resize_test_engine_pool!(size::Int64, generator::Option{Function}=nothing)
+function resize_test_engine_pool!(size::Int64, name_generator::Option{Function}=nothing)
     if size < 0
         size = 0
     end
 
-    if !isnothing(generator)
-        TEST_ENGINE_POOL.generator = generator
+    if !isnothing(name_generator)
+        TEST_ENGINE_POOL.name_generator = name_generator
     end
 
+    @lock TEST_SERVER_LOCK begin
+        # Add engines if size > length
+        _create_and_add_engines!(size)
+        _validate_engine_pool!()
+        _trim_engine_pool!(size)
+    end
+end
+
+# Test all engines and remove if they are unavailable or not successfully provisioned
+function _validate_engine_pool!()
+    @lock TEST_SERVER_LOCK begin
+        @sync for engine in TEST_ENGINE_POOL.engines
+            try
+                response = get_engine(get_context(), engine.first; readtimeout=30)
+                if response.state == "PROVISIONED"
+                    # Success! Move on and try the next engine
+                    continue
+                end
+                # The engine exists, but is not provisioned despite our best attempts
+            catch
+                # The engine does not exist
```

Comment:
404 means the engine does not exist. There are plans for a 429 with a Retry-After, although that shouldn't be handled here. Then there are HTTP connection errors that are not about the engine.

Reporting the error, whatever it is, could be useful. 404 is a definitive error, but handling that would be covered by reporting the error. The response, if there is one, could also be reported.

Reply:
but it sounds like we do expect those to be `HTTPError` and not some other error? Maybe check for that just to avoid a situation where we make a typo in the try block and swallow it

Reply:
Ah, right. Only `HTTPError`s make sense for us. If we do get a different error, something is seriously wrong and we should rethrow.